Document Ubuntu Core image creation without static kernel cmdline params (#56)
canonical/real-time-ubuntu-docs: docs/how-to/uc-image-creation.rst

    This guide assumes access to features available exclusively to `dedicated Snap Store`_ users.

Canonical builds and publishes Ubuntu Core images for a range of `supported platforms`_.
However, these images include the generic Linux kernel. 
In order to run Ubuntu Core with the real-time kernel, we need to build it ourselves. 

> This guide shows how to build and Ubuntu Core image with the real-time kernel.

Reviewer: ```suggestion
This guide shows how to build an Ubuntu Core image with the real-time kernel.
```

We start by showing how to :ref:`build a vanilla image <ubuntu-core-image>` which is excellent for testing and tuning, and then illustrate how to :ref:`make a custom image <custom-ubuntu-core-image>` with production-ready kernel configurations.

.. _ubuntu-core-image:

Create an image with the real-time kernel
-----------------------------------------

To do this we need to describe the image content and then use a tool to build it.

.. _model-assertion:

Create the model assertion
~~~~~~~~~~~~~~~~~~~~~~~~~~

The model assertion is a digitally signed document that describes the content of the Ubuntu Core image.
Read the `model assertion`_ documentation before continuing.

Below is an example model assertion in YAML, describing a `core22` Ubuntu Core
image:

.. literalinclude :: uc-image-creation/model.json
   :language: json

Inside an empty directory, create a file named ``model.json`` with the above content.

Change the following:

- ``authority-id``, ``brand-id`` to your developer ID, since this is custom model. Use ``snapcraft whoami`` command to get your developer ID.

> - ``timestamp`` to a RFC3339 formatted string, within you signing key's validity. Use ``date -Iseconds --utc`` command to use the current time.

Author: Please let me know if this helps: https://github.com/canonical/real-time-ubuntu-docs/pull/56/commits/83f6139f4ec8e4c7064c9aaef1b3d1cd84565018

- ``store`` to your dedicated Snap Store ID.
- ``model`` to something representative of your model.

The ``snaps`` array is a list of snaps that get included in the image.
In that list, the ``realtime-kernel`` snap contains the realtime Linux kernel.
Here you can add any other snaps, including for example your real-time applications.

Next, we need to sign the model assertion.
Refer to the guide on `signing model assertion`_ for details on how to sign the model assertion. 

Here are the needed steps:

1) Create and register a key

.. code-block:: shell

    snapcraft create-key realtime-ubuntu

>     snapcraft register-key realtime-ubuntu

Reviewer: Regarding this, it's important to advise the user to choose the key **name** and **password** carefully because once registered, [it's not possible to remove the key from the Snapcraft store](https://forum.snapcraft.io/t/remove-registered-signing-key-from-snap-store/27280).

Consider adding a warning in this section.

Author: Not sure why they need to choose the name and password carefully. Based on the explanation on the forum post, the registered key is the public part and can't be removed.

Reviewer: Let me put this differently. It’s not so much about being careful, but rather being **aware** of the implications of creating a key. This process is **permanent** and **irreversible**, and anything with those characteristics should come with a warning to the user. 

Additionally, the user cannot delete a key even if it isn't registered yet. Since creating a key requires performing `snapcraft login` first, the key remains tied to the user account regardless.

The key reasons to warn the user are:

1. **Irreversibility**: Permanent actions, by definition, cannot be undone. The user may not realize they are performing an irreversible action.

2. **Security Implications**: The user is creating a key with a password, and they won’t be able to change the password after the key is created and registered.

3. **Avoiding Future Frustration**: The user could become frustrated if, after following the documentation, they wish to clean up everything and discover they can't.

This warning is important to prevent user mistakes, provide a better user experience, and ensure clarity when following the documentation.

Author: > Additionally, the user cannot delete a key even if it isn't registered yet. Since creating a key requires performing snapcraft login first, the key remains tied to the user account regardless.

I don't think that's true. The key is stored locally. Only registered public keys are uploaded and tied to an Ubuntu account.

Please provide your feedback to snapcraft and core documentations as they are out of the scope of this document. As mentioned, the user is asked to read the documentation before starting.

> 

Reviewer: Here, you might want to add a brief explanation:

- Creating a key refers to generating a GPG key.
- Registering the key means uploading the public GPG key to the Snapcraft Store account.

Author: I'd leave this to the respective documentation mentioned right above:

> Refer to the guide on [signing model assertion](https://ubuntu.com/core/docs/sign-model-assertion) for details on how to sign the model assertion. 
> 
> Here are the needed steps:

> 

Reviewer: ```suggestion

.. warning::

    Choose the name and password for the key carefully, as once registered, it's not possible to change or remove the key.
```

Reviewer: > We don't know when, but once that happens,

Not sure if that will ever happen. If they didn't implement key deletion at the time they implemented key registration, there must be another reason for it.

Author: This is a presumption on someone else's product. A bit of research shows that the snapcraft CLI used to support revocation, and there are locally available revocation certificates that can be used for each key. 

https://bugs.launchpad.net/snapstore-server/+bug/1669471

```console
$ tree /home/farshid/.snap/gnupg/
/home/farshid/.snap/gnupg/
├── openpgp-revocs.d
│   ├── 32C84846EF48EF6BD8F97CC63DDF5FF093D7F9B2.rev
│   ├── 5D2F0F60249592D1A49CB80BA8C4F62D7912B49E.rev
│   ├── 9142937D8F532836CC5D23E713E7CFC9CCED927E.rev
│   └── 93F56C15D58577B69DADB02DC92BA06F835B494F.rev
├── private-keys-v1.d
│   ├── 2F138999CDDB4A3C9730D681624AF2534DCFD7E8.key
│   ├── 8060443C480D86672B610D4B70E3BB6D662E784F.key
│   └── E2814AF986655128FC204D0FF044B1790EB315C7.key
├── pubring.kbx
├── pubring.kbx~
├── sshcontrol
└── trustdb.gpg
```

>     snapcraft create-key realtime-ubuntu

Author: I changed this and extended the explanations: https://github.com/canonical/real-time-ubuntu-docs/pull/56/commits/625f6481e7864841ecc4668617acb97535ec4860

Then build the image:

.. code-block:: console

    $ UBUNTU_STORE_AUTH_DATA_FILENAME=credentials.txt ubuntu-image snap model.signed.yaml --verbose --validation=enforce
    [0] prepare_image
    Fetching snapd (21759)
    Fetching realtime-kernel (149)
    Fetching core22 (1586)
    Fetching pc (146)
    [1] load_gadget_yaml
    [2] set_artifact_names
    [3] populate_rootfs_contents
    [4] generate_disk_info
    [5] calculate_rootfs_size
    [6] populate_bootfs_contents
    [7] populate_prepare_partitions
    [8] make_disk
    [9] generate_snap_manifest
    Build successful

This downloads all the snaps specified in the model assertion and builds an image file called ``pc.img``.

.. hint::

    To fetch the ``realtime-kernel`` snap for this image build, it should be included explicitly in your dedicated Snap Store.

.. code-block:: console

    $ file pc.img 
    pc.img: DOS/MBR boot sector; partition 1 : ID=0xee, start-CHS (0x0,0,0), end-CHS (0x0,0,0), startsector 1, 6195199 sectors, extended partition table (last)

✅ The image file is now ready to be flashed on a medium to create a bootable drive with the Ubuntu Core installer!

----

> At this point, you could continue by tuning your system for real-time processing. 

Author: I changed the sentence but prefer to keep the horizontal line to add emphasis on this. I would leave this for @AnneCYH to decide, during her review on #45.

    2 directories, 2 files

The project directory should contain the model assertion, the signed model assertion, and the pc-gadget directory.

Build the Ubuntu Core image
---------------------------

> Similar to before, we use ``ubuntu-image`` but this time we need to also provide the path to the custom gadget snap file:

Reviewer 2: ```suggestion
Similar to before, we use ``ubuntu-image`` to build the image. This time we also need to provide the path to the custom gadget snap file. You therefore need:
```

    [9] generate_snap_manifest
    Build successful

This adds all the snaps specified in the model assertion and builds an image file called ``pc.img``.
There is a warning for ``realtime-pc`` gadget snap because this is being side-loaded, rather than fetched from the store.

✅ The image file with the custom configurations is ready! 

> On a running instance based off this image, the kernel parameters can be verified by looking into ``/proc/cmdline``:

Reviewer 2: ```suggestion
After installing and running a device with this image, the kernel parameters can be verified by looking into ``/proc/cmdline``:
```
